contracts/utils/ShortStrings.sol

```diff
@@ -0,0 +1,72 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity ^0.8.8;
+
+import "./StorageSlot.sol";
+
+/**
+ * @dev Short string operations.
+ */
+library ShortStrings {
+    type ShortString is bytes32;
+
+    error StringTooLong(string str);
+
+    /**
+     * @dev Encode a string of at most 31 chars into a `ShortString`.
+     *
+     * This will trigger a `StringTooLong` error is the input string is too long.
+     */
+    function toShortString(string memory str) internal pure returns (ShortString) {
+        bytes memory bstr = bytes(str);
+        if (bstr.length > 31) {
+            revert StringTooLong(str);
+        }
+        return ShortString.wrap(bytes32(uint256(bytes32(bstr)) | bstr.length));
+    }
+
+    /**
+     * @dev Decode a `ShortString` back to a "normal" string.
+     */
+    function toString(ShortString sstr) internal pure returns (string memory) {
+        uint256 len = length(sstr);
+        // using `new string(len)` would work locally but is not memory safe.
+        string memory str = new string(32);
+        /// @solidity memory-safe-assembly
+        assembly {
+            mstore(str, len)
+            mstore(add(str, 0x20), sstr)
+        }
+        return str;
+    }
+
+    /**
+     * @dev Return the length of a `ShortString`.
+     */
+    function length(ShortString sstr) internal pure returns (uint256) {
+        return uint256(ShortString.unwrap(sstr)) & 0xFF;
+    }
+
+    /**
+     * @dev Encode a string into a `ShortString`, or write it to storage if it is too long.
+     */
+    function setWithFallback(string memory value, string storage store) internal returns (ShortString) {
```

Comment:
The "fallback" is storage, so I don't like `setWithFallback` as a name because you only "set" when you fallback.

What about `toShortStringWithFallback`?

Similarly we could use `toStringWithFallback`

Reply:
I was thinking

```solidity
contract NamedContract {
    ShortStrings.ShortString private _name;
    string private _nameFallback;
    
    constructor(string memory name_) {
        _name = ShortStrings.setWithFallback(name_, _nameFallback);
    }
    
    function name() public view returns (string memory) {
        return ShortStrings.getWithFallback(_name, _nameFallback);
    }
}
```

How would you do it?

Reply:
would you prefer

```
constructor(string memory name_) {
    _name = name_.toShortStringWithFallback(_nameFallback);
}
function name() public view returns (string memory) {
    return _name.toStringWithFallback(, _nameFallback);
}
```
?

Reply:
I was thinking the same. Those are good.

We should add that as an example snippet in the docs.